File: src/09.tree/39.449.go
package tree

import (
	"strconv"
	"strings"
)

// Serializes a tree to a single string.
func Serialize449(root *TreeNode) string {
	if root == nil {
		return ""
	}

	return strconv.Itoa(root.Val) + " " + Serialize449(root.Left) + Serialize449(root.Right)
}

// Deserializes your encoded data to tree.
func Deserialize449(data string) *TreeNode {
	if data == "" {
		return nil
	}
	var root *TreeNode = nil
	var insert func(node *TreeNode, val int) *TreeNode
	insert = func(node *TreeNode, val int) *TreeNode {
		if node == nil {
			return &TreeNode{Val: val}
		} else if val > node.Val {
			node.Right = insert(node.Right, val)
		} else {
			node.Left = insert(node.Left, val)
		}

		return node
	}

	data = data[:len(data)-1]
	arr := strings.Split(data, " ")

	for _, v := range arr {
		c, _ := strconv.Atoi(v)
		root = insert(root, c)
	}

	return root
}

File: src/09.tree/39.449.ts
import { TreeNode } from "./01.100";

/*
 * Encodes a tree to a single string.
 */
export function serialize449(root: TreeNode | null): string {
    if (!root) {
        return "";
    }
    return root.val + " " + serialize449(root.left) + serialize449(root.right);
}

/*
 * Decodes your encoded data to tree.
 */
export function deserialize449(data: string): TreeNode | null {
    if (!data.length) {
        return null;
    }

    data = data.slice(0, -1);

    const arr = data.split(" ");
    const insert = (node: TreeNode | null, val: number): TreeNode | null => {
        if (!node) {
            return new TreeNode(val);
        } else if (val > node.val) {
            node.right = insert(node.right, val);
        } else {
            node.left = insert(node.left, val);
        }

        return node;
    };

    let root: TreeNode | null = null;
    arr.forEach((item) => {
        const v = Number(item);
        root = insert(root, v);
    });

    return root;
}
